### DISCUSSION TITLE
Формирование списка из отмеченных галочками строк и вывод на печать.

### DISCUSSION THREAD

**Discussion by Kmetyan** [totumonline/totum-ru-issues-and-discussions#24] [Как это сделать?] [answered]
Приветствую!
Есть таблица, необходимо взять значения 2 столбцов таблицы из строк отмеченных галочками и вывести на печать!

Поле код действия содержит следующее:

= : linkToPrint(template: "Reestr_sertif"; data: $row1)
row1: rowCreate(field: "reestr" = $example)
example: select(table: 'uchet_vydachi_sertifikatov'; field: 'seriynye_nomera'; where: 'id' = $num1)
num1: var(name: "count"; value: $#ids)

Скрин: 
![image](https://user-images.githubusercontent.com/91886195/136326941-ca6f655e-c834-4814-bca3-dfdc4798779b.png)

На печать выводит только одно значение (с наименьшим ID) хотя выделено 3 строки.
Скрин: 
![image](https://user-images.githubusercontent.com/91886195/136327468-dcee8dd4-e14c-421c-98ab-3909ef366e78.png)

Подскажите как сделать что бы выводились все значение на печать?

**Comment by ttm-alexeypolunin** [comment, marked as answer]
О, мне нравятся такие вопросы!

Тотум жестко разделяет одинарные значения и множественные — они же списки или list.

Соответсвенно функции имеют два варианта — одинарные (например [select](https://docs.totum.online/functions#fn-select)) и множественные ([selectList](https://docs.totum.online/functions#fn-selectList)).

В коде, который вы привели в строке `example` используется **одинарный вариант (`select`)**, поэтому берется одно значение с наименьшим `id` (если не задан `order` то не обязательно так — это определяет `postgres` у себя в кишках).

Я модифицирую код на множественный:

```totum
= : linkToPrint(template: "Reestr_sertif"; data: $row1)
row1: rowCreate(field: "reestr" = $example)
example: selectList(table: 'uchet_vydachi_sertifikatov'; field: 'seriynye_nomera'; where: 'id' = $#ids)
```
Еще я убрал `var` и прописал `where: 'id' = $#ids` — тк тут не нужна дополнительная строка кода.

**Сделав это, вы столкнетесь с ошибкой!** тк они `list`, а шаблон печати ожидает `text` *(он может и list но это отдельный разговор про вложенные шаблоны).*

Добавим превращение `list` —> `text` функцией [listJoin](https://docs.totum.online/functions#fn-listJoin):

```totum
= : linkToPrint(template: "Reestr_sertif"; data: $row1)
row1: rowCreate(field: "reestr" = $example)
example: listJoin(list: $list; str: "<br>")
list: selectList(table: 'uchet_vydachi_sertifikatov'; field: 'seriynye_nomera'; where: 'id' = $#ids)
```
В качестве разделителя используем `<br>`, а не быструю переменную `$#nl` (это сокращение от `new line`) тк `html` не понимает эти разделители линий.

Теперь уже должно работать нормально, но будет косяк, когда галочки не стоят (ни одна). Добавим в начало проверку наличия галочек. Переменная `$#ids` это список (list), поэтому нам надо сравнить ее с пустым `list` — `$#lc`:

```totum
= : if(condition: $#ids != $#lc; then: $print; else: $err)
err: linkToDataText(title: "Предупреждение печати"; text: "Выберите хотя бы одну позицию")
print: linkToPrint(template: "Reestr_sertif"; data: $row1)
row1: rowCreate(field: "reestr" = $example)
example: listJoin(list: $list; str: "<br>")
list: selectList(table: 'uchet_vydachi_sertifikatov'; field: 'seriynye_nomera'; where: 'id' = $#ids)
```

Вот обычно в таком варианте это используется в наших проектах.

**Comment by Kmetyan** [comment]
Очень помогли, все заработало. Спасибо.
